jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -332,29 +372,105 @@ public String toString()
+         * another thread.  If the caller wishes to acquire the associated entry, they should
+         * use {@link #acquire(Object)} to atomically enable and acquire.
+         * @param pooled The pooled item for the entry
+         */
+        public void enable(T pooled)
+        {
+            enable(pooled, false);
+        }
+
+        /** Enable and acquire the reserved {@link Entry}.
+         * The associated entry is atomically enabled and acquired, so that no other thread can acquire it and
+         * the {@link #getEntry()} value may be used by the caller.
+         * @param pooled The pooled item for the entry
+         */
+        public Pool<T>.Entry acquire(T pooled)
+        {
+            enable(pooled, true);
+            return entry;
+        }
+
+        private void enable(T pooled, boolean acquire)
+        {
+            Objects.requireNonNull(pooled);
+            if (entry.state.getHi() != Integer.MIN_VALUE)
+                throw new IllegalStateException("Open entries cannot be enabled : " + this);
+            entry.pooled = pooled;
+            int usage = acquire ? 1 : 0;
+            if (!entry.state.compareAndSet(Integer.MIN_VALUE, usage, 0, usage))
+            {
+                entry.pooled = null;
+                throw new IllegalStateException("Entry cannot be enabled : " + this);
+            }
+            pending.decrementAndGet();
+        }
+
+        /**
+         * Remove the reservation without enabling.
+         */
+        public void remove()
+        {
+            Pool.this.remove(entry);
+        }
+
+        @Override
+        public String toString()
+        {
+            return String.format("%s@%x{%s}",
+                getClass().getName(),
+                hashCode(),
+                entry);
+        }
+    }
+
     public class Entry
     {
-        // hi: positive=open/maxUsage counter,negative=closed lo: multiplexing counter
+        // hi: positive=open/maxUsage counter; negative=closed; MIN_VALUE pending
+        // lo: multiplexing counter
         private final AtomicBiInteger state;
-        private volatile T pooled;
 
-        public Entry()
+        // The pooled item.  This is not volatile as it is set once and then never changed.
+        // Other threads accessing must check the state field above first, so a good before/after
+        // relationship exists to make a memory barrier.
+        private T pooled;
+
+        Entry()
         {
-            this.state = new AtomicBiInteger(-1, 0);
+            this.state = new AtomicBiInteger(Integer.MIN_VALUE, 0);
         }
 
         public T getPooled()
         {
             return pooled;
         }
 
-        public void enable(T pooled)
+        /**
+         * Release the entry.
+         * This is equivalent to calling {@link Pool#release(Pool.Entry)} passing this entry.
+         * @return true if released.
+         */
+        public boolean release()
```

Comment:
This method is never used, so remove it.

Reply:
It will be once I do my XmlConfiguration PR.
It makes sense that an item in the pool that has a reference to it's own entry can release itself without needing to keep a reference to the pool.

Reply:
I'm not sure. `Entry` is not really an API class now that you have introduced `Reservation`.
It's an opaque mechanism that needs to be exposed, but I think it's more correct to force people to use `Pool` and `Reservation`and `Attachable` rather than `Entry`.

Reply:
`Entry` most definitely is an API class and I have code that pools objects that remember their own `Entry`.  It is pointless to force them to also remember the pool they came from just so they can call release.... as the `Entry` already knows its own pool.